File: SDK Examples/Examples/Java/src/AO_SynchronousOneWaveformAO/SynchronousOneWaveformAO.java
package AO_SynchronousOneWaveformAO;

import java.awt.Color;
import java.awt.EventQueue;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.WindowAdapter;
import java.awt.event.WindowEvent;
import java.text.NumberFormat;

import javax.swing.ImageIcon;
import javax.swing.JButton;
import javax.swing.JFormattedTextField;
import javax.swing.JFrame;
import javax.swing.JLabel;
import javax.swing.JOptionPane;
import javax.swing.JPanel;
import javax.swing.SwingConstants;
import javax.swing.UIManager;
import javax.swing.border.EmptyBorder;

import Automation.BDaq.*;
import Common.*;

import org.eclipse.wb.swing.FocusTraversalOnArray;

import java.awt.Component;

public class SynchronousOneWaveformAO extends JFrame {

	/**
	 * Define the default serialization number.
	 */
	private static final long serialVersionUID = 1L;

	private JPanel contentPane;
	
	private JFormattedTextField txtHighLevelA;
	private JFormattedTextField txtLowLevelA;
	private JFormattedTextField txtHighLevelB;
	private JFormattedTextField txtLowLevelB;
	private JButton btnSineA;
	private JButton btnSquareA;
	private JButton btnTriangleA;
	private JButton btnSineB;
	private JButton btnSquareB;
	private JButton btnTriangleB;
	private JButton btnConfigure;
	private JButton btnStart;
	private JLabel lblChannelA;
	private JLabel lblChannelB;
	private ImageIcon[][] imageIcon = {
			{
				new BackgroundPanel("sine.png","sine").getImageIcon(),
				new BackgroundPanel("square.png","square").getImageIcon(),
				new BackgroundPanel("triangle.png","triangle").getImageIcon()
			},
			{
				new BackgroundPanel("sine_down.png","sine").getImageIcon(),
				new BackgroundPanel("square_down.png","square").getImageIcon(),
				new BackgroundPanel("triangle_down.png","triangle").getImageIcon()
			}
	}; 
	
	private BufferedAoCtrl bufferedAoCtrl = new BufferedAoCtrl();
	private ConfigureDialog configureDialog;
	ConfigureParameter configure = new ConfigureParameter();
	private double[] dataScaled ;
	private WaveformGenerator waveformGenerator;
	
	private double highLevelA = 0;
	private double lowLevelA = 0;
	private double highLevelB = 0;
	private double lowLevelB = 0;
	
	private WaveformStyle waveStyleA; 
	private WaveformStyle waveStyleB;
	
	private int[][] buttonState = {{0, 0, 0},{0, 0, 0}};
	private boolean isChannelAChecked = false;
	private boolean isChannelBChecked = false;

	/**
	 * 
	 * Build Date:2011-9-6 
	 * Author:Administrator
	 * Function Description: Launch the application.
	 */
	public static void main(String[] args) {
		EventQueue.invokeLater(new Runnable() {
			public void run() {
				try {
					SynchronousOneWaveformAO frame = new SynchronousOneWaveformAO();
					frame.setVisible(true);
				} catch (Exception e) {
					e.printStackTrace();
				}
			}
		});
	}

	/**
	 * 
	 * Build Date:2011-9-6 
	 * Author:Administrator
	 * Function Description: Create the frame.
	 */
	public SynchronousOneWaveformAO() {
		// Add window close action listener.
		addWindowListener(new WindowCloseActionListener());
		
		setTitle("Synchonous One Waveform AO - Run");
		setResizable(false);
		setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
		setBounds(100, 100, 355, 470);
		contentPane = new JPanel();
		contentPane.setBorder(new EmptyBorder(5, 5, 5, 5));
		setContentPane(contentPane);
		contentPane.setLayout(null);

		// Add background image for the main Frame
		BackgroundPanel panel = new BackgroundPanel("AO1.png", "background image!");
		panel.setBounds(0, 0, 348, 438);
		contentPane.add(panel);
		panel.setLayout(null);

		txtHighLevelA = new JFormattedTextField(NumberFormat.getNumberInstance());
		txtHighLevelA.setValue(new Double(5));
		txtHighLevelA.setHorizontalAlignment(SwingConstants.RIGHT);
		txtHighLevelA.setBounds(82, 89, 75, 22);
		panel.add(txtHighLevelA);
		txtHighLevelA.setColumns(10);

		txtLowLevelA = new JFormattedTextField(NumberFormat.getNumberInstance());
		txtLowLevelA.setValue(new Double(-5));
		txtLowLevelA.setHorizontalAlignment(SwingConstants.RIGHT);
		txtLowLevelA.setColumns(10);
		txtLowLevelA.setBounds(192, 88, 75, 22);
		panel.add(txtLowLevelA);

		btnSineA = new JButton(imageIcon[0][0]);
		btnSineA.addActionListener(new WaveButtonActionListener());
		btnSineA.setBounds(80, 121, 53, 50);
		panel.add(btnSineA);

		btnSquareA = new JButton(imageIcon[0][1]);
		btnSquareA.addActionListener(new WaveButtonActionListener());
		btnSquareA.setBounds(149, 121, 53, 50);
		panel.add(btnSquareA);

		btnTriangleA = new JButton(imageIcon[0][2]);
		btnTriangleA.addActionListener(new WaveButtonActionListener());
		btnTriangleA.setBounds(216, 121, 53, 50);
		panel.add(btnTriangleA);

		btnSineB = new JButton(imageIcon[0][0]);
		btnSineB.addActionListener(new WaveButtonActionListener());
		btnSineB.setBounds(82, 308, 53, 50);
		panel.add(btnSineB);

		btnSquareB = new JButton(imageIcon[0][1]);
		btnSquareB.addActionListener(new WaveButtonActionListener());
		btnSquareB.setBounds(149, 308, 53, 50);
		panel.add(btnSquareB);

		btnTriangleB = new JButton(imageIcon[0][2]);
		btnTriangleB.addActionListener(new WaveButtonActionListener());
		btnTriangleB.setBounds(214, 308, 53, 50);
		panel.add(btnTriangleB);

		txtHighLevelB = new JFormattedTextField(NumberFormat.getNumberInstance());
		txtHighLevelB.setValue(new Double(5));
		txtHighLevelB.setHorizontalAlignment(SwingConstants.RIGHT);
		txtHighLevelB.setColumns(10);
		txtHighLevelB.setBounds(82, 274, 75, 22);
		panel.add(txtHighLevelB);

		txtLowLevelB = new JFormattedTextField(NumberFormat.getNumberInstance());
		txtLowLevelB.setValue(new Double(-5));
		txtLowLevelB.setHorizontalAlignment(SwingConstants.RIGHT);
		txtLowLevelB.setColumns(10);
		txtLowLevelB.setBounds(192, 274, 75, 22);
		panel.add(txtLowLevelB);

		btnConfigure = new JButton("Configure");
		btnConfigure.setBounds(61, 395, 115, 23);
		btnConfigure.addActionListener(new ConfigureButtonActionListener());
		panel.add(btnConfigure);
		
		lblChannelA = new JLabel("0");
		lblChannelA.setForeground(new Color(188, 143, 143));
		lblChannelA.setFont(UIManager.getFont("InternalFrame.titleFont"));
		lblChannelA.setBounds(230, 33, 26, 15);
		panel.add(lblChannelA);
		
		lblChannelB = new JLabel("1");
		lblChannelB.setForeground(new Color(188, 143, 143));
		lblChannelB.setFont(UIManager.getFont("InternalFrame.titleFont"));
		lblChannelB.setBounds(230, 219, 26, 15);
		panel.add(lblChannelB);
		
		btnStart = new JButton("Start");
		btnStart.setBounds(209, 395, 90, 23);
		btnStart.addActionListener(new StartButtonActionListener());
		panel.add(btnStart);
		panel.setFocusTraversalPolicy(new FocusTraversalOnArray(new Component[]{
				txtHighLevelA, txtLowLevelA, btnSineA, btnSquareA, btnTriangleA, 
				txtHighLevelB, txtLowLevelB, btnSineB, btnSquareB, btnTriangleB, 
				btnConfigure, btnStart}));
		
		configureDialog  = new ConfigureDialog(this);
		configureDialog.setModal(true);
		configureDialog.setVisible(true);
	}
	
	/**
	 * 
	 * Build Date:2011-9-6 
	 * Author:Administrator
	 * Function Description: This function is used to configure the device of bufferedAoCtrl.
	 */
	protected void ConfigureDevice() {
		try {
			bufferedAoCtrl.setSelectedDevice(new DeviceInformation(configure.deviceName));
			String profilePath = null;
			profilePath = configureDialog.GetProfilePath();
			ErrorCode errorCode = bufferedAoCtrl.LoadProfile(profilePath);
			if(Global.BioFaild(errorCode)){
				ShowMessage("Sorry, there're some errors occred, ErrorCode: " + errorCode.toString());
				}
			
			AnalogChannel[] channels = bufferedAoCtrl.getChannels();
			bufferedAoCtrl.getScanChannel().setChannelStart(configure.channelStart);
			bufferedAoCtrl.getScanChannel().setChannelCount(configure.channelCount);
			bufferedAoCtrl.getScanChannel().setSamples(configure.pointCountPerWave);
			bufferedAoCtrl.getScanChannel().setIntervalCount(configure.pointCountPerWave);
			bufferedAoCtrl.getConvertClock().setRate(configure.clockRate);
			
			for (int i = 0; i < channels.length; i++) {
				channels[i].setValueRange(configure.valueRange);
			}	
		} catch (Exception e) {
			ShowMessage("Sorry, there're some errors occred: " + e.getMessage());
			return;
		}
		ErrorCode errorCode = bufferedAoCtrl.Prepare();
		if(Global.BioFaild(errorCode)){
			ShowMessage("Sorry, there're some errors occred, ErrorCode: " + errorCode.toString());
		}
	}
	
	/**
	 * 
	 *Build Date:2011-9-6
	 *Author:Administrator
	 *Function Description: This function is used to initialize the wave button.
	 */
	private void InitializeWaveButton(){
		for(int i = 0; i < 2; i++){
			for(int j = 0; j < 3; j++){
				if(j == 0){
					buttonState[i][j] = 1;
				}else{
					buttonState[i][j] = 0;
				}		
			}
		}
		btnSineA.setIcon(imageIcon[1][0]);
		btnSquareA.setIcon(imageIcon[0][1]);
		btnTriangleA.setIcon(imageIcon[0][2]);
		waveStyleA = WaveformStyle.sine;
		isChannelAChecked = true;
		
		btnSineB.setIcon(imageIcon[1][0]);
		btnSquareB.setIcon(imageIcon[0][1]);
		btnTriangleB.setIcon(imageIcon[0][2]);
		waveStyleB = WaveformStyle.sine;
		isChannelBChecked = true;
	}
	
	/**
	 * 
	 *Build Date:2011-9-6
	 *Author:Administrator
	 *Function Description: This function is used to set the initialized state of the main frame.
	 */
	private void ConfigurePanel(){
		AoFeatures feature = bufferedAoCtrl.getFeatures();
		
		InitializeWaveButton();
		lblChannelA.setText(String.valueOf(configure.channelStart));
		
		if(configure.channelCount == 1){
			lblChannelB.setText("");
			btnSineB.setEnabled(false);
			btnSquareB.setEnabled(false);
			btnTriangleB.setEnabled(false);
			txtHighLevelB.setEnabled(false);
			txtLowLevelB.setEnabled(false);
		}else{
			int chanNumberB = configure.channelStart + 1;
			if(chanNumberB == feature.getChannelCountMax()){
				chanNumberB = 0;
			}
			lblChannelB.setText(String.valueOf(chanNumberB));
			btnSineB.setEnabled(true);
			btnSquareB.setEnabled(true);
			btnTriangleB.setEnabled(true);
			txtHighLevelB.setEnabled(true);
			txtLowLevelB.setEnabled(true);
		}
	}
	
	/**
	 * 
	 * Build Date:2011-9-6 
	 * Author:Administrator
	 * Function Description: This function is used to initialize the main frame.
	 */
	public void Initialization(){
		ConfigureDevice();
		
		//'2 * pointCountPerWave' means we handle two channels in this demo at most. 
		int rawDataBufferLength = 2 * configure.pointCountPerWave;
		waveformGenerator = new WaveformGenerator(configure.pointCountPerWave);
		
		if(dataScaled == null || dataScaled.length < rawDataBufferLength){
			dataScaled = new double[rawDataBufferLength];
		}
		
		this.setTitle("Synchonous One Waveform AO - Run(" + configure.deviceName.substring(0, configure.deviceName.length() - 1) + ")");
		
		ConfigurePanel();
	}
	
	/**
	 * 
	 * Build Date:2011-9-6
	 * Author:Administrator
	 * Function Description: This function is used to show the error massage to user!
	 * 
	 * @param message: the message shown to users!
	 */
	protected void ShowMessage(String message) {
		JOptionPane.showMessageDialog(this, message, "Warning MessageBox",
				JOptionPane.WARNING_MESSAGE);
	}
	
	/**
	 * 
	 * @author Administrator
	 * Class Function Description: This class is used to listen the waveform button 's action.
	 */
	class WaveButtonActionListener implements ActionListener {
		public void actionPerformed(ActionEvent e) {
			JButton source = (JButton)e.getSource();
			if(source == btnSineA){
				if(buttonState[0][0] == 1){
					buttonState[0][0] = 0;
					isChannelAChecked = false;
				}else{
					buttonState[0][0] = 1;
					waveStyleA = WaveformStyle.sine;
					isChannelAChecked = true;
				}
				buttonState[0][1] = 0;
				buttonState[0][2] = 0;
			}else if(source == btnSquareA){
				if(buttonState[0][1] == 1){
					buttonState[0][1] = 0;
					isChannelAChecked = false;
				}else{
					buttonState[0][1] = 1;
					waveStyleA = WaveformStyle.square;
					isChannelAChecked = true;
				}
				buttonState[0][0] = 0;
				buttonState[0][2] = 0;
			}else if(source == btnTriangleA){
				if(buttonState[0][2] == 1){
					buttonState[0][2] = 0;
					isChannelAChecked = false;
				}else{
					buttonState[0][2] = 1;
					waveStyleA = WaveformStyle.triangle;
					isChannelAChecked = true;
				}
				buttonState[0][0] = 0;
				buttonState[0][1] = 0;
			}else if(source == btnSineB){
				if(buttonState[1][0] == 1){
					buttonState[1][0] = 0;
					isChannelBChecked = false;
				}else{
					buttonState[1][0] = 1;
					waveStyleB = WaveformStyle.sine;
					isChannelBChecked = true;
				}
				buttonState[1][1] = 0;
				buttonState[1][2] = 0;
			}else if(source == btnSquareB){
				if(buttonState[1][1] == 1){
					buttonState[1][1] = 0;
					isChannelBChecked = false;
				}else{
					buttonState[1][1] = 1;
					waveStyleB = WaveformStyle.square;
					isChannelBChecked = true;
				}
				buttonState[1][0] = 0;
				buttonState[1][2] = 0;
			}else if(source == btnTriangleB){
				if(buttonState[1][2] == 1){
					buttonState[1][2] = 0;
					isChannelBChecked = false;
				}else{
					buttonState[1][2] = 1;
					waveStyleB = WaveformStyle.triangle;
					isChannelBChecked = true;
				}
				buttonState[1][0] = 0;
				buttonState[1][1] = 0;
			}
			
			if(buttonState[0][0] == 1){
				btnSineA.setIcon(imageIcon[1][0]);
			}else{
				btnSineA.setIcon(imageIcon[0][0]);
			}
			
			if(buttonState[0][1] == 1){
				btnSquareA.setIcon(imageIcon[1][1]);
			}else{
				btnSquareA.setIcon(imageIcon[0][1]);
			}
			
			if(buttonState[0][2] == 1){
				btnTriangleA.setIcon(imageIcon[1][2]);
			}else{
				btnTriangleA.setIcon(imageIcon[0][2]);
			}
			
			if(buttonState[1][0] == 1){
				btnSineB.setIcon(imageIcon[1][0]);
			}else{
				btnSineB.setIcon(imageIcon[0][0]);
			}
			
			if(buttonState[1][1] == 1){
				btnSquareB.setIcon(imageIcon[1][1]);
			}else{
				btnSquareB.setIcon(imageIcon[0][1]);
			}
			
			if(buttonState[1][2] == 1){
				btnTriangleB.setIcon(imageIcon[1][2]);
			}else{
				btnTriangleB.setIcon(imageIcon[0][2]);
			}
		}
	}
	
	/**
	 * 
	 * @author Administrator
	 * Class Function Description: This class is used to listen the configure button 's action.
	 */
	class ConfigureButtonActionListener implements ActionListener {
		public void actionPerformed(ActionEvent e) {
			configureDialog.isFirstLoad = false;
			configureDialog.setVisible(true);
		}
	}
	
	/**
	 * 
	 * @author Administrator
	 * Class Function Description: This class is used to listen the configure button 's action.
	 */
	class StartButtonActionListener implements ActionListener {
		public void actionPerformed(ActionEvent e) {
			btnStart.setEnabled(false);
			btnConfigure.setEnabled(false);

			if (!isChannelAChecked && !isChannelBChecked) {
				ShowMessage("Please select a waveform !");
				return;
			}
			
			highLevelA = ((Number)txtHighLevelA.getValue()).doubleValue();
			lowLevelA = ((Number)txtLowLevelA.getValue()).doubleValue();
			highLevelB = ((Number)txtHighLevelB.getValue()).doubleValue();
			lowLevelB = ((Number)txtLowLevelB.getValue()).doubleValue();
			
			WaveformStyle[] styles = new WaveformStyle[] { waveStyleA, waveStyleB };
			double[] highLevels = new double[] { highLevelA, highLevelB };
			double[] lowLevels = new double[] { lowLevelA, lowLevelB };
			if (configure.channelCount <= 1) {
				if (isChannelAChecked) {
					waveformGenerator.GetOnePeriod(dataScaled, waveStyleA, highLevelA, lowLevelA);
				} else {
					ShowMessage("Please select one waveform style!");
					return;
				}

				ErrorCode errorCode = bufferedAoCtrl.SetData(configure.pointCountPerWave, dataScaled);
				if (Global.BioFaild(errorCode)) {
					ShowMessage("Sorry, there're some errors occred, ErrorCode: "
							+ errorCode.toString());
					return;
				}

			} else {
				if (isChannelAChecked && !isChannelBChecked) {
					styles[0] = waveStyleA;
					highLevels[1] = 0;
					lowLevels[1] = 0;
				} else if (!isChannelAChecked && isChannelBChecked) {
					styles[1] = waveStyleB;
					highLevels[0] = 0;
					lowLevels[0] = 0;
				} else if (!isChannelAChecked && !isChannelBChecked) {
					highLevels[0] = 0;
					highLevels[1] = 0;
					lowLevels[0] = 0;
					lowLevels[1] = 0;
				}

				waveformGenerator.GetMultipleWaveform(dataScaled, 2, styles, highLevels, lowLevels);
				ErrorCode errorCode = bufferedAoCtrl.SetData(2 * configure.pointCountPerWave,
						dataScaled);
				if (Global.BioFaild(errorCode)) {
					ShowMessage("Sorry,there're some errors occurred,ErrorCode: " 
							+ errorCode.toString());
					return;
				}
			}
			
			ErrorCode errorCode = bufferedAoCtrl.RunOnce();
			if (Global.BioFaild(errorCode)) {
				ShowMessage("Sorry, there're some errors occred, ErrorCode: "
						+ errorCode.toString());
				return;
			}

			btnStart.setEnabled(true);
			btnConfigure.setEnabled(true);
		}
	}
	
	/**
	 * 
	 * @author Administrator
	 * Class Function Description: This class is used to listen the main frame's closing event.
	 */
	class WindowCloseActionListener extends WindowAdapter{
		@Override
		public void windowClosing(WindowEvent e) {
			if (bufferedAoCtrl != null) {
				bufferedAoCtrl.Cleanup();
			}
		}
	}
}

/**
 * 
 * @author Administrator 
 * Class Description: this class is use to transfer parameter to the main frame.
 */
class ConfigureParameter{
	public String deviceName;
	public int channelStart;
	public int channelCount;
	public ValueRange valueRange;
	public int pointCountPerWave;
	public double clockRate;
}

File: SDK Examples/Examples/Qt/AO_SynchronousOneWaveformAO/synchronousonewaveformao.cpp
#include "synchronousonewaveformao.h"
#include <QMessageBox>

SynchronousOneWaveformAO::SynchronousOneWaveformAO(QWidget *parent, Qt::WindowFlags flags)
	: QDialog(parent, flags)
{
	ui.setupUi(this);
	
	//Set the minimum and close button of the main frame.
	this->setWindowFlags(Qt::WindowFlags(Qt::WindowSystemMenuHint | Qt::WindowTitleHint | Qt::WindowCloseButtonHint));

   bfdAoCtrl = BufferedAoCtrl::Create();

   dataScaled = NULL;
	this->m_waveSeled[0] = false;
	this->m_waveSeled[1] = false;

	this->buttonGroup1 = new QButtonGroup();
	this->buttonGroup1->addButton(ui.BtnSineA, 0);
	this->buttonGroup1->addButton(ui.BtnTriangleA, 1);
	this->buttonGroup1->addButton(ui.BtnSquareA, 2);
	this->buttonGroup1->addButton(ui.BtnSineB, 3);
	this->buttonGroup1->addButton(ui.BtnTriangleB, 4);
	this->buttonGroup1->addButton(ui.BtnSquareB, 5);
	this->buttonGroup1->setExclusive(false);

	this->buttons[0] = ui.BtnSineA;
	this->buttons[1] = ui.BtnTriangleA;
	this->buttons[2] = ui.BtnSquareA;
	this->buttons[3] = ui.BtnSineB;
	this->buttons[4] = ui.BtnTriangleB;
	this->buttons[5] = ui.BtnSquareB;
	
	strs[0] = "background:url(:/SynchronousOneWaveformAO/Resources/sine.png)";
	strs[1] = "background:url(:/SynchronousOneWaveformAO/Resources/sine_down.png)";
	strs[2] = "background:url(:/SynchronousOneWaveformAO/Resources/triangle.png)";
	strs[3] = "background:url(:/SynchronousOneWaveformAO/Resources/triangle_down.png)";
	strs[4] = "background:url(:/SynchronousOneWaveformAO/Resources/square.png)";
	strs[5] = "background:url(:/SynchronousOneWaveformAO/Resources/square_down.png)";

	connect(ui.btnConfigure, SIGNAL(clicked()), this, SLOT(ButtonConfigureClicked()));
	connect(ui.btnStart, SIGNAL(clicked()), this, SLOT(ButtonStartClicked()));
	connect(buttonGroup1, SIGNAL(buttonClicked(int)), this, SLOT(WaveButtonClicked(int)));
}

SynchronousOneWaveformAO::~SynchronousOneWaveformAO()
{
	
}

void SynchronousOneWaveformAO::Initialize() {
	//set the title of the form.
	this->setWindowTitle(tr("Synchronous One Waveform AO - Run(") + configure.deviceName + tr(")"));
	
	if (dataScaled != NULL) {
        delete[] dataScaled;
        dataScaled = NULL;
    }

	waveformGenerator = new WaveformGenerator(configure.pointCountPerWave);

	InitialButtons();

	//initialize the sine as the default choice
	ui.BtnSineA->setChecked(true);
	ui.BtnSineA->setStyleSheet(strs[1]);
	m_waveSeled[0] = true;
	m_waveParam[0].Style = sine;

	if (configure.channelCount > 1) {
		ui.BtnSineB->setChecked(true);
		ui.BtnSineB->setStyleSheet(strs[1]);
		m_waveSeled[1] = true;
		m_waveParam[1].Style = sine;
	}

	ConfigureDevice();

	//set the ChannelCount
    if (bfdAoCtrl->getScanChannel()->getChannelCount() > 1) {
        bfdAoCtrl->getScanChannel()->setChannelCount(2);
    }

	ConfigurePanel();
}

void SynchronousOneWaveformAO::ConfigureDevice() {
	ErrorCode errorCode = Success;
    std::wstring description = configure.deviceName.toStdWString();
    DeviceInformation selected(description.c_str());
	errorCode = bfdAoCtrl->setSelectedDevice(selected);
	CheckError(errorCode);
	errorCode = bfdAoCtrl->LoadProfile(configure.profilePath);
	CheckError(errorCode);

	errorCode = bfdAoCtrl->setStreaming(false);
    CheckError(errorCode);

	ScanChannel* scanChannel = bfdAoCtrl->getScanChannel();
    errorCode = scanChannel->setChannelStart(configure.channelStart);
    CheckError(errorCode);
    errorCode = scanChannel->setChannelCount(configure.channelCount);
    CheckError(errorCode);
    errorCode = scanChannel->setSamples(configure.pointCountPerWave);
    CheckError(errorCode);
    errorCode = bfdAoCtrl->getConvertClock()->setRate(configure.clockRate);
    CheckError(errorCode);

    for (int i = 0; i < bfdAoCtrl->getChannels()->getCount(); i++) {
        errorCode = bfdAoCtrl->getChannels()->getItem(i).setValueRange(configure.valueRange);
        CheckError(errorCode);
    }
}

void SynchronousOneWaveformAO::CheckError(ErrorCode errorCode) {
    if (BioFailed(errorCode)) {
		QString message = tr("Sorry, there are some errors occurred, Error Code: 0x") +
		QString::number(errorCode, 16).right(8).toUpper();
		QMessageBox::information(this, "Warning Information", message);
	}
}

void SynchronousOneWaveformAO::ButtonConfigureClicked() {
	int dialogResult = configDialog->exec();
	if (dialogResult == QDialog::Accepted) {
		configure = configDialog->GetConfigureParameter();
		Initialize();
	}
}

void SynchronousOneWaveformAO::ConfigurePanel() {
	if (configure.channelCount == 1) {
		ui.BtnSineB->setEnabled(false);
		ui.BtnTriangleB->setEnabled(false);
		ui.BtnSquareB->setEnabled(false);
		ui.txtboxHiLevelB->setEnabled(false);
		ui.txtboxLoLevelB->setEnabled(false);
		ui.chLabelB->setText(tr(""));
	} else {
		ui.BtnSineB->setEnabled(true);
		ui.BtnTriangleB->setEnabled(true);
		ui.BtnSquareB->setEnabled(true);
		ui.txtboxHiLevelB->setEnabled(true);
		ui.txtboxLoLevelB->setEnabled(true);
		ui.chLabelB->setText(QString("%1").arg((configure.channelStart + 1) % bfdAoCtrl->getChannelCount()));
	}
	ui.chLabelA->setText(QString("%1").arg(configure.channelStart));
}

void SynchronousOneWaveformAO::InitialButtons() {
	for (int i = 0; i < 6; i++) {
		buttons[i]->setChecked(false);
		buttons[i]->setStyleSheet(strs[(i % 3) * 2]);
	}
}

void SynchronousOneWaveformAO::WaveButtonClicked(int id) {
	int seledCH, baseIdx, imgIdx;

	seledCH = id / 3;
	baseIdx = seledCH * 3;

	for (int i = baseIdx; i < baseIdx + 3; i++) {
		if (i != id && buttons[i]->isChecked()) {
			imgIdx = (i- baseIdx) * 2;
			buttons[i]->setChecked(false);
			buttons[i]->setStyleSheet(strs[imgIdx]);
		}
	}

	if (buttons[id]->isChecked()) {
		buttons[id]->setStyleSheet(strs[(id - baseIdx) * 2 + 1]);
	} else {
		buttons[id]->setStyleSheet(strs[(id - baseIdx) * 2]);
	}

	m_waveSeled[seledCH] = buttons[id]->isChecked();

	if (buttons[id]->isChecked()) {
		m_waveParam[seledCH].Style = (WaveformStyle)(id % 3);
	} else {
		m_waveParam[seledCH].Style = (WaveformStyle)2;
	}
}

void SynchronousOneWaveformAO::ButtonStartClicked() {
	int dataCount;
	ErrorCode errorCode = Success;

	if ((m_waveSeled[0] == false) && (m_waveSeled[1] == false)) {
		QMessageBox::information(this, "Warning Information", "Please select a waveform!");
		return;
	}

	//prepare the AO
    errorCode = bfdAoCtrl->Prepare();
    CheckError(errorCode);

	ScanChannel* scanChannel = bfdAoCtrl->getScanChannel();
    dataCount = scanChannel->getChannelCount() * scanChannel->getSamples();
	dataScaled = new double[dataCount];

	if ((m_waveSeled[0] == true) && (m_waveSeled[1] == false)) {
		m_waveParam[0].HighLevel = ui.txtboxHiLevelA->text().toDouble();
        m_waveParam[1].HighLevel = 0;
		m_waveParam[0].LowLevel = ui.txtboxLoLevelA->text().toDouble();
        m_waveParam[1].LowLevel = 0;
    } else if ((m_waveSeled[0] == false) && (m_waveSeled[1] == true)) {
        m_waveParam[0].HighLevel = 0;
		m_waveParam[1].HighLevel = ui.txtboxHiLevelB->text().toDouble();
        m_waveParam[0].LowLevel = 0;
		m_waveParam[1].LowLevel = ui.txtboxLoLevelB->text().toDouble();
    } else if ((m_waveSeled[0] == true) && (m_waveSeled[1] == true)) {
        m_waveParam[0].HighLevel = ui.txtboxHiLevelA->text().toDouble();
        m_waveParam[1].HighLevel = ui.txtboxHiLevelB->text().toDouble();
        m_waveParam[0].LowLevel = ui.txtboxLoLevelA->text().toDouble();
        m_waveParam[1].LowLevel = ui.txtboxLoLevelB->text().toDouble();
    }

	//generator the wave
	if (configure.channelCount <= 1) {
		this->waveformGenerator->GetOnePeriod(dataScaled, m_waveParam[0]);
	} else {
		this->waveformGenerator->GetMultipleWaveform(dataScaled, 2, m_waveParam);
	}

	//Convert scaled  data to raw data
	errorCode = bfdAoCtrl->SetData(dataCount, dataScaled);
    CheckError(errorCode);

    //Start buffered AO in synchronous
	errorCode = bfdAoCtrl->RunOnce();
    CheckError(errorCode);
}
